### PR TITLE
Relax set_num_threads restriction in parallel native case

### PR DESCRIPTION
**Pull request by ilia-cher**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27925 Add CI builds for TBB and native parallel builds
* **#27947 Relax set_num_threads restriction in parallel native case**

Summary:
Don't throw exception if the requested size is the same as the currently
used one

Test Plan:
ATEN_THREADING=NATIVE python setup.py develop --cmake

Reviewers:liujiakai

Differential Revision: [D17919416](https://our.internmc.facebook.com/intern/diff/D17919416)

### CODE REVIEW

File: aten/src/ATen/ParallelNative.cpp [hunk 1]
```diff
@@ -178,10 +178,19 @@ void set_num_threads(int nthreads) {
 #ifndef C10_MOBILE
   TORCH_CHECK(nthreads > 0, "Expected positive number of threads");
   int no_value = NOT_SET;
-  TORCH_CHECK(num_intraop_threads.compare_exchange_strong(no_value, nthreads),
-      "Error: cannot set number of intraop threads "
-      "after parallel work has started or after set_num_threads call "
-      "when using native parallel backend");
+  if (!num_intraop_threads.compare_exchange_strong(no_value, nthreads)) {
```
Comment (dskhudia) [inline aten/src/ATen/ParallelNative.cpp]: So now calling, for example, set_num_threads(2) twice doesn't raise exception but set_num_threads(2) and then again set_num_threads(4) still raises exception with native thread pool. Correct?

**Review by dskhudia — APPROVED** [review]

File: aten/src/ATen/ParallelNative.cpp (hunk 1 above)
Comment (ilia-cher) [reply aten/src/ATen/ParallelNative.cpp]: yes, this is correct. The main reason for this is that we actually do have places that call set_num_threads(x) multiple times (where x is usually 1), and as of now, we may assume this to be a supported behavior. Changing number of threads though is not supported in native backend

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by ilia-cher** [comment]
using warning instead of exception

**Comment by ilia-cher** [comment]
Regarding throwing exception vs. showing warning, it seems a bunch of tests are still broken, because of some use of set_num_threads, although real use cases should be fine afaik. We might try just showing a warning instead of throwing exception

**Pull request closed without merging**